File: ml-backend/cpl_offline_kit/pipeline/quality.ts
/**
 * CPL Crop Doctor — Offline Image Quality Assessment
 *
 * Direct TypeScript port of: cpl_crop.validation.quality.py
 *
 * Runs cheap, deterministic checks on an image BEFORE expensive
 * segmentation + classification. Catches blurry / dark / overexposed /
 * tiny images so we reject them with helpful retake guidance.
 *
 * USAGE IN EXPO:
 *   This module operates on raw RGBA pixel data from the camera.
 *   Use expo-image-manipulator or react-native-image-crop-picker to
 *   get image pixel data, then convert to a grayscale Uint8Array.
 *
 *   Alternatively, if you're using a canvas (e.g., react-native-canvas),
 *   you can call getImageData() and pass the RGBA data directly.
 */

import type { QualityReport } from "./types";

// ─── Default Thresholds (matching Python config.py) ──────────────────

export const QUALITY_DEFAULTS = {
  minResolution: 224,
  minBlur: 100.0,
  minBrightness: 40.0,
  maxBrightness: 220.0,
  minContrast: 25.0,
} as const;

// ─── Helper: clamp a value to [0, 1] with linear ramp ────────────────

function normalizeToUnit(value: number, lo: number, hi: number): number {
  if (hi <= lo) return value >= hi ? 1.0 : 0.0;
  return Math.max(0.0, Math.min(1.0, (value - lo) / (hi - lo)));
}

// ─── Core: Convert RGBA pixel buffer to grayscale Uint8Array ─────────

/**
 * Convert an RGBA Uint8Array (from canvas getImageData or camera)
 * to a single-channel grayscale Uint8Array.
 *
 * Formula: gray = 0.299*R + 0.587*G + 0.114*B (standard luminance)
 */
export function rgbaToGrayscale(
  rgba: Uint8Array | Uint8ClampedArray,
  width: number,
  height: number
): Uint8Array {
  const gray = new Uint8Array(width * height);
  for (let i = 0; i < width * height; i++) {
    const r = rgba[i * 4];
    const g = rgba[i * 4 + 1];
    const b = rgba[i * 4 + 2];
    gray[i] = Math.round(0.299 * r + 0.587 * g + 0.114 * b);
  }
  return gray;
}

// ─── Blur: Laplacian variance approximation ──────────────────────────

/**
 * Compute an approximate Laplacian variance (sharpness metric).
 *
 * This is the variance of a 3x3 Laplacian kernel convolved over the
 * grayscale image. Higher values = sharper image. The Python version
 * uses cv2.Laplacian(gray, cv2.CV_64F).var() — this is equivalent.
 *
 * Kernel:  [0  1  0]
 *          [1 -4  1]
 *          [0  1  0]
 */
function laplacianVariance(
  gray: Uint8Array,
  width: number,
  height: number
): number {
  let sum = 0;
  let sumSq = 0;
  let count = 0;

  // Skip 1px border (kernel needs neighbors)
  for (let y = 1; y < height - 1; y++) {
    for (let x = 1; x < width - 1; x++) {
      const idx = y * width + x;
      const lap =
        gray[idx - width] + // top
        gray[idx - 1] + // left
        -4 * gray[idx] + // center
        gray[idx + 1] + // right
        gray[idx + width]; // bottom

      sum += lap;
      sumSq += lap * lap;
      count++;
    }
  }

  if (count === 0) return 0;
  const mean = sum / count;
  return sumSq / count - mean * mean; // variance
}

// ─── Main: Assess image quality ──────────────────────────────────────

export interface QualityOptions {
  minResolution?: number;
  minBlur?: number;
  minBrightness?: number;
  maxBrightness?: number;
  minContrast?: number;
}

/**
 * Compute a quality report for an image.
 *
 * @param gray - Grayscale Uint8Array (use rgbaToGrayscale to convert)
 * @param width - Image width in pixels
 * @param height - Image height in pixels
 * @param opts - Optional threshold overrides
 * @returns QualityReport with ok, score, and individual metrics
 */
export function assessImageQuality(
  gray: Uint8Array,
  width: number,
  height: number,
  opts?: QualityOptions
): QualityReport {
  const minRes = opts?.minResolution ?? QUALITY_DEFAULTS.minResolution;
  const minBlur = opts?.minBlur ?? QUALITY_DEFAULTS.minBlur;
  const minBright = opts?.minBrightness ?? QUALITY_DEFAULTS.minBrightness;
  const maxBright = opts?.maxBrightness ?? QUALITY_DEFAULTS.maxBrightness;
  const minContrast = opts?.minContrast ?? QUALITY_DEFAULTS.minContrast;

  // --- Compute metrics ---

  const blur = laplacianVariance(gray, width, height);

  // Brightness = mean grayscale value
  let brightnessSum = 0;
  for (let i = 0; i < gray.length; i++) brightnessSum += gray[i];
  const brightness = brightnessSum / gray.length;

  // Contrast = standard deviation of grayscale values
  let varianceSum = 0;
  for (let i = 0; i < gray.length; i++) {
    const diff = gray[i] - brightness;
    varianceSum += diff * diff;
  }
  const contrast = Math.sqrt(varianceSum / gray.length);

  // --- Hard fail checks ---

  const failures: string[] = [];
  if (Math.min(height, width) < minRes) {
    failures.push(`resolution_below_${minRes}`);
  }
  if (blur < minBlur) {
    failures.push(`blurry_below_${minBlur}`);
  }
  if (brightness < minBright) {
    failures.push(`too_dark_below_${minBright}`);
  } else if (brightness > maxBright) {
    failures.push(`too_bright_above_${maxBright}`);
  }
  if (contrast < minContrast) {
    failures.push(`low_contrast_below_${minContrast}`);
  }

  // --- Soft sub-scores in [0, 1] ---

  const resScore = normalizeToUnit(Math.min(height, width), minRes, minRes * 2);
  const blurScore = normalizeToUnit(blur, minBlur, minBlur * 4);

  let brightScore: number;
  if (brightness >= minBright && brightness <= maxBright) {
    brightScore = 1.0;
  } else if (brightness < minBright) {
    brightScore = normalizeToUnit(brightness, 0, minBright);
  } else {
    brightScore = normalizeToUnit(255 - brightness, 0, 255 - maxBright);
  }

  const contrastScore = normalizeToUnit(contrast, minContrast, minContrast * 2);

  const score = (resScore + blurScore + brightScore + contrastScore) / 4;

  return {
    ok: failures.length === 0,
    score,
    resolution: [height, width],
    blur,
    brightness,
    contrast,
    failures,
  };
}

File: ml-backend/hf_deploy_stage/src/cpl_crop/validation/quality.py
"""Stage 3 — OpenCV-based image-quality assessment.

Cheap, deterministic checks that run *before* expensive segmentation +
classification. Catches blurry / dark / overexposed / tiny images so
we can reject them with helpful retake guidance.

All thresholds are config-driven via :class:`cpl_crop.config.Settings`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

import cv2
import numpy as np
from numpy.typing import NDArray


@dataclass(frozen=True)
class QualityReport:
    """Result of :func:`assess_image_quality`. All numeric fields are populated."""

    ok: bool
    score: float  # 0..1, where 1 = all checks pass cleanly
    resolution: tuple[int, int]  # (height, width)
    blur: float  # Laplacian variance — higher = sharper
    brightness: float  # 0..255 mean grayscale
    contrast: float  # std of grayscale pixels
    leaf_ratio: float  # proportion of leaf-colored pixels (green/yellow/brown)
    failures: list[str] = field(default_factory=list)

    def to_json(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "score": round(self.score, 4),
            "resolution": list(self.resolution),
            "blur": round(self.blur, 2),
            "brightness": round(self.brightness, 2),
            "contrast": round(self.contrast, 2),
            "leaf_ratio": round(self.leaf_ratio, 4),
            "failures": list(self.failures),
        }


def _normalize_to_unit(value: float, lo: float, hi: float) -> float:
    """Linear ramp clamped to [0, 1]."""
    if hi <= lo:
        return 1.0 if value >= hi else 0.0
    return float(max(0.0, min(1.0, (value - lo) / (hi - lo))))


def assess_image_quality(
    image_rgb: NDArray[np.uint8],
    *,
    min_resolution: int = 224,
    min_blur: float = 100.0,
    min_brightness: float = 40.0,
    max_brightness: float = 220.0,
    min_contrast: float = 25.0,
    min_leaf_ratio: float = 0.05,  # At least 5% of the image must be leaf-colored
) -> QualityReport:
    """Compute a quality report for an RGB uint8 image.

    The overall ``score`` is a soft average of four sub-scores so the
    confidence engine can use it directly. ``ok`` is the AND of hard
    pass/fail predicates.
    """
    if image_rgb.ndim != 3 or image_rgb.shape[-1] != 3:
        raise ValueError(f"Expected (H, W, 3) RGB image; got shape {image_rgb.shape}")

    h, w = image_rgb.shape[:2]
    gray = cv2.cvtColor(image_rgb, cv2.COLOR_RGB2GRAY)

    blur = float(cv2.Laplacian(gray, cv2.CV_64F).var())
    brightness = float(gray.mean())
    contrast = float(gray.std())

    # --- Leaf Color (Green/Yellow/Brown) Detection ---
    hsv = cv2.cvtColor(image_rgb, cv2.COLOR_RGB2HSV)
    # Hue: 10 (brown/orange) to 85 (green)
    # Saturation: > 30 (ignore grays/whites)
    # Value: > 30 (ignore dark blacks)
    lower_leaf = np.array([10, 30, 30])
    upper_leaf = np.array([85, 255, 255])
    leaf_mask = cv2.inRange(hsv, lower_leaf, upper_leaf)
    leaf_ratio = float(cv2.countNonZero(leaf_mask)) / (h * w)

    failures: list[str] = []
    if min(h, w) < min_resolution:
        failures.append(f"resolution_below_{min_resolution}")
    if blur < min_blur:
        failures.append(f"blurry_below_{min_blur:.0f}")
    if brightness < min_brightness:
        failures.append(f"too_dark_below_{min_brightness:.0f}")
    elif brightness > max_brightness:
        failures.append(f"too_bright_above_{max_brightness:.0f}")
    if contrast < min_contrast:
        failures.append(f"low_contrast_below_{min_contrast:.0f}")
    if leaf_ratio < min_leaf_ratio:
        failures.append(f"not_a_leaf_ratio_{leaf_ratio:.2f}")

    # Soft sub-scores in [0, 1]
    res_score = _normalize_to_unit(min(h, w), min_resolution, min_resolution * 2)
    blur_score = _normalize_to_unit(blur, min_blur, min_blur * 4)
    if min_brightness <= brightness <= max_brightness:
        bright_score = 1.0
    elif brightness < min_brightness:
        bright_score = _normalize_to_unit(brightness, 0.0, min_brightness)
    else:
        bright_score = _normalize_to_unit(255.0 - brightness, 0.0, 255.0 - max_brightness)
    contrast_score = _normalize_to_unit(contrast, min_contrast, min_contrast * 2)
    leaf_score = _normalize_to_unit(leaf_ratio, min_leaf_ratio, min_leaf_ratio * 3)

    score = float(np.mean([res_score, blur_score, bright_score, contrast_score, leaf_score]))

    return QualityReport(
        ok=not failures,
        score=score,
        resolution=(h, w),
        blur=blur,
        brightness=brightness,
        contrast=contrast,
        leaf_ratio=leaf_ratio,
        failures=failures,
    )
